### PR TITLE
Track nullability information in publicAPI files for C# compiler

### PR DESCRIPTION
**Pull request by jcouv**
Note: this PR was updated to track APIs involving oblivious reference types with a `~` prefix.

This PR ran into issue https://github.com/dotnet/roslyn/issues/50357
But the public API fixer can be run with the instructions below:

1.	Install dotnet-format as a global tool. It does ship as part of the SDK, but a separate version can be installed as a global tool and invoked with dotnet-format {options}.
`C:\Source\roslyn> dotnet tool install -g dotnet-format --version "6.*" --add-source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json`
2.	Restore and build Compilers.sln. This is necessary to ensure the source generator project is built and we can load the generator assembly when running dotnet-format.
`C:\Source\roslyn> .\restore.cmd`
`C:\Source\roslyn> .\build.cmd`
3.	Invoke the dotnet-format global tool. Running only the analyzers subcommand and fixing only the “missing Public API signature” diagnostic. We must also pass the --include-generated flag to include source generated documents in the analysis.
`C:\Source\roslyn> cd ..`
`C:\Source> dotnet-format analyzers .\roslyn\Compilers.sln --diagnostics=RS0016 --no-restore --include-generated -v diag`

### CODE REVIEW

**Comment by jcouv** [comment]
Here's the latest set of oblivious APIs:

```
Code
~Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax.Update(Microsoft.CodeAnalysis.SyntaxToken asyncKeyword, Microsoft.CodeAnalysis.SyntaxToken delegateKeyword, Microsoft.CodeAnalysis.CSharp.Syntax.ParameterListSyntax parameterList, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode body) -> Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax.WithAsyncKeyword(Microsoft.CodeAnalysis.SyntaxToken asyncKeyword) -> Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax.WithBody(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode body) -> Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.LambdaExpressionSyntax.WithAsyncKeyword(Microsoft.CodeAnalysis.SyntaxToken asyncKeyword) -> Microsoft.CodeAnalysis.CSharp.Syntax.LambdaExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.LambdaExpressionSyntax.WithBody(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode body) -> Microsoft.CodeAnalysis.CSharp.Syntax.LambdaExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax.Update(Microsoft.CodeAnalysis.SyntaxToken asyncKeyword, Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax parameter, Microsoft.CodeAnalysis.SyntaxToken arrowToken, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode body) -> Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax.WithAsyncKeyword(Microsoft.CodeAnalysis.SyntaxToken asyncKeyword) -> Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax
~Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax.WithBody(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode body) -> Microsoft.CodeAnalysis.CSharp.Syntax.SimpleLambdaExpressionSyntax
~static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.AnonymousMethodExpression() -> Microsoft.CodeAnalysis.CSharp.Syntax.AnonymousMethodExpressionSyntax

~Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader.AddDependencyLocation(string fullPath) -> void
~Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader.LoadFromPath(string fullPath) -> System.Reflection.Assembly
~Microsoft.CodeAnalysis.ISymbol.ContainingAssembly.get -> Microsoft.CodeAnalysis.IAssemblySymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingModule.get -> Microsoft.CodeAnalysis.IModuleSymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingNamespace.get -> Microsoft.CodeAnalysis.INamespaceSymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingSymbol.get -> Microsoft.CodeAnalysis.ISymbol
~Microsoft.CodeAnalysis.ISymbol.ContainingType.get -> Microsoft.CodeAnalysis.INamedTypeSymbol
~Microsoft.CodeAnalysis.SeparatedSyntaxList<TNode>.GetEnumerator() -> Microsoft.CodeAnalysis.SeparatedSyntaxList<TNode>.Enumerator
~Microsoft.CodeAnalysis.SyntaxList<TNode>.GetEnumerator() -> Microsoft.CodeAnalysis.SyntaxList<TNode>.Enumerator
```

**Comment by jcouv** [comment]
I'll try to push this PR forward using `dotnet forward` once https://github.com/dotnet/roslyn/issues/51608 is merged.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Thanks

File: src/Compilers/CSharp/Portable/Syntax/AnonymousMethodExpressionSyntax.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 #nullable disable
+#pragma warning disable RS0041 // uses oblivious reference types
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/AnonymousMethodExpressionSyntax.cs:6]: There aren't that many APIs in these files. Should we just annotate them?

**Comment by 333fred** [comment]
Done review pass (commit 1)

File: src/Compilers/CSharp/Portable/Syntax/AnonymousMethodExpressionSyntax.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/AnonymousMethodExpressionSyntax.cs:6]: I'd tried that but it snowballed the change. Admittedly this was a while back, but my recollection is some of those have many usage sites that needed adjustment. So I'd kept the PR focused on its primary purpose.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Syntax/AnonymousMethodExpressionSyntax.cs:6]: Good enough reason for me.

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Could you please include the instructions in the PR description in some in-repo documentation?

**Comment by jcouv** [comment]
> Could you please include the instructions in the PR description in some in-repo documentation?

Done. However the instructions will need to be updated/simplified when certain bits are published. @JoeRobich can advise.

**Review by RikkiGibson — APPROVED** [review]
LGTM, but it still wasn't clear to me how/when the fixer was changed to include non-annotated and oblivious reference types.

**Comment by AlekseyTs** [comment]
In my opinion, the PR title "Update Public API files" includes no useful information that one generally expects to find there.

**Comment by jcouv** [comment]
@RikkiGibson The functionality for the publicAPI analyzer/fixer to track nullability annotations was added a [while back](https://github.com/dotnet/roslyn-analyzers/pulls?q=is%3Apr+author%3Ajcouv+is%3Aclosed). We've turned on `#nullable enable` in the Core project, but couldn't do it for the C# project, because the fixer doesn't work well with source-generators. The instructions are a mitigation for that.

**Pull request merged**